### PR TITLE
feat: bsc and matic integration

### LINKED ISSUE #2886: [Demand] Mask on BSC Tech Preview
## Catalogue

- [ ] UI/UX
- [x] Mask Core Feature
- [ ] Mask Plugin Feature
- [ ] Security/Cryptography Feature
- [ ] Other
  - /_List it here_/

### LINKED ISSUE #3028: [Demand] Deploy contracts on BSC
## Catalogue

- [ ] UI/UX
- [ ] Mask Core Feature
- [x] Mask Plugin Feature
- [ ] Security/Cryptography Feature
- [ ] Other
  - /_List it here_/

## Part 2: Problem to Solve

+ [ ] RedPacket Contract
+ [ ] ITO Contract
+ [ ] ITO Qualification Contract

In the beginning, only deploy to the testnet of BSC.

## Notes

+ https://chainid.network/

### PR DESCRIPTION
**Pull request by guanbinrui**
closes #2886
closes #3028

### CODE REVIEW

File: packages/maskbook/src/plugins/Wallet/UI/ImportWalletDialog/index.tsx [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+    const [words, setWords] = useState<string[]>(BLANK_WORDS)
+    const tabState = useState(0)
+    const tabs: AbstractTabProps['tabs'] = useMemo(
+        () => [
+            {
+                label: 'Mnemonic',
+                children: <MnemonicTab words={words} onChange={setWords} />,
+            },
+            {
+                label: 'Json File',
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Wallet/UI/ImportWalletDialog/index.tsx:81]:
```suggestion
                label: 'JSON File',
```

File: packages/maskbook/src/plugins/Wallet/UI/ImportWalletDialog/index.tsx [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+    const [walletFromPrivateKey, setWalletFromPrivateKey] = useState<Partial<RecoverResult>>({})
+
+    const { open, closeDialog } = useRemoteControlledDialog(WalletMessages.events.importWalletDialogUpdated)
+
+    const [words, setWords] = useState<string[]>(BLANK_WORDS)
+    const tabState = useState(0)
+    const tabs: AbstractTabProps['tabs'] = useMemo(
+        () => [
+            {
+                label: 'Mnemonic',
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Wallet/UI/ImportWalletDialog/index.tsx:77]:
```suggestion
                label: 'Mnemonic' Words,
```

File: packages/maskbook/src/_locales/en/messages.json [hunk 3]
```diff
@@ -306,27 +307,27 @@
...
     "plugin_wallet_wrong_network_tip": "Please connect to the appropriate Ethereum network.",
+    "plugin_wallet_recent_transaction": "Recent Transactions",
+    "plugin_wallet_clear_all": "Clear All",
     "plugin_wallet_on_create": "Create Wallet",
     "plugin_wallet_on_connect": "Connect Wallet",
     "plugin_wallet_setup_title_create": "Create a Wallet",
-    "plugin_wallet_setup_description_words": "Your secret backup phrase makes it easy to back up and restore your account. Write this phrase on a piece of paper and store it in a secure location. If you want even more security, write it down on multiple pieces of paper and store each in 2 - 3 different locations.",
+    "plugin_wallet_create_import_choose": "Create Wallet & Import Wallet",
+    "plugin_wallet_create_a_new_wallet": "Create a New Wallet",
+    "plugin_wallet_create_a_new_wallet_description": "Mask network supports Eth and Polka networks ",
```
Comment (guanbinrui) [inline packages/maskbook/src/_locales/en/messages.json:317]: Inconsistent with the facts.

File: packages/maskbook/src/plugins/Wallet/services/recentTransactions.ts [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+import { RecentTransactionRecordIntoDB } from './helpers'
+
+const MAX_RECENT_TRANSACTIONS_SIZE = 5
+
+export async function getRecentTransactions(address: string) {
+    const t = createTransaction(await createWalletDBAccess(), 'readonly')('Transaction')
+    const transactions = await t.objectStore('Transaction').getAll()
+    return transactions
+        .filter((x) => isSameAddress(x.address, address))
+        .sort((a, z) => a.createdAt.getTime() - z.createdAt.getTime())
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Wallet/services/recentTransactions.ts:16]: Should order it in reverse direction.

File: packages/maskbook/src/_locales/en/messages.json [hunk 5]
```diff
@@ -343,6 +344,7 @@
     "plugin_wallet_settings_portfolio_data_source_secondary": "Select the source of portfolio data.",
     "plugin_wallet_settings_collectible_data_source_primary": "Collectible Data Source",
     "plugin_wallet_settings_collectible_data_source_secondary": "Select the source of collectible data.",
+    "plugin_wallet_name_placeholder": "Enter 1-12 characters",
```
Comment (septs) [inline packages/maskbook/src/_locales/en/messages.json:347]:
```suggestion
    "plugin_wallet_name_placeholder": "Enter max 12 characters (required)",
```

**Comment by nuanyang233** [comment]
Please resolve eslint warning

File: packages/maskbook/src/plugins/Wallet/UI/ConnectWalletDialog/index.tsx [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+                ],
+            })
+            return true as const
+        } catch (e) {
+            throw new Error(e.message)
+        }
+    }, [account, chainId, selectedNetworkType, selectedProviderType])
+
+    const connectToWalletConnect = useCallback(async () => {
+        const [uri_] = await Promise.allSettled([await Services.Ethereum.createConnectionURI(), delay(1000)])
```
Comment (nuanyang233) [inline packages/maskbook/src/plugins/Wallet/UI/ConnectWalletDialog/index.tsx:89]: The 1000 looks like a magic number. 👀

File: packages/maskbook/src/plugins/Wallet/UI/CreateWalletDialog.tsx [hunk 7]
```diff
@@ -217,9 +223,34 @@ export function CreateWalletDialog(props: CreateWalletDialogProps) {
...
+                            />
+                        }
+                        label={
+                            <Box
+                                sx={{
+                                    display: 'inline-flex',
+                                    alignItems: 'flex-start',
+                                }}>
+                                <Typography className={classes.confirmation} variant="body2">
+                                    I have securely written down my mnemonic word, I understand that lost mnemonic word
```
Comment (nuanyang233) [inline packages/maskbook/src/plugins/Wallet/UI/CreateWalletDialog.tsx:242]: Perhaps need i18n

File: packages/maskbook/src/plugins/Wallet/UI/CreateWalletDialog.tsx [hunk 8]
```diff
@@ -231,18 +262,22 @@ export function CreateWalletDialog(props: CreateWalletDialogProps) {
...
+                        {step === CreateWalletStep.NameAndWords
                             ? t('plugin_wallet_setup_create')
-                            : step === CreateWalletStep.Words
-                            ? t('plugin_wallet_setup_next')
                             : t('plugin_wallet_setup_verify')}
                     </ActionButton>
                 </Box>
+                {step === CreateWalletStep.NameAndWords && (
+                    <Box className={classes.warning}>
+                        <Alert severity="warning">Please properly back up your account’s mnemonic words.</Alert>
```
Comment (nuanyang233) [inline packages/maskbook/src/plugins/Wallet/UI/CreateWalletDialog.tsx:278]: Perhaps need i18n

File: packages/maskbook/src/plugins/Wallet/UI/ImportWalletDialog/FromJson.tsx [hunk 9]
```diff
@@ -0,0 +1,5 @@
+import type { FC } from 'react'
+
+export const FromJson: FC = () => {
+    return <div>from json</div>
```
Comment (nuanyang233) [inline packages/maskbook/src/plugins/Wallet/UI/ImportWalletDialog/FromJson.tsx:4]: Emm，Why is such a component needed, as this now seems pointless

File: packages/maskbook/src/plugins/Wallet/UI/ImportWalletDialog/FromPrivateKey.tsx [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+        }, 500)()
+    }, [privateKey])
+
+    return (
+        <Box>
+            <TextField
+                className={classes.textField}
+                multiline
+                error={!valid}
+                helperText={!valid ? 'The Private key is incorrect' : ''}
```
Comment (nuanyang233) [inline packages/maskbook/src/plugins/Wallet/UI/ImportWalletDialog/FromPrivateKey.tsx:66]: Perhaps need i18n

File: packages/maskbook/src/plugins/Wallet/UI/ImportWalletDialog/ImportResult.tsx [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+        <>
+            <Typography variant="h1" className={classes.title}>
+                Derivation path (m'/ 44'/ 60'/ 0 /'0 )
+            </Typography>
+            <Table>
+                <TableHead>
+                    <TableRow>
+                        <TableCell className={classes.headCell}>Address</TableCell>
+                        <TableCell align="center" className={classes.headCell}>
+                            Balances(ETH)
```
Comment (nuanyang233) [inline packages/maskbook/src/plugins/Wallet/UI/ImportWalletDialog/ImportResult.tsx:68]: Perhaps need i18n

File: packages/maskbook/src/plugins/Wallet/UI/ImportWalletDialog/ImportResult.tsx [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+        borderBottom: 'none',
+        backgroundColor: '#F3F3F4',
+    },
+    bodyCell: {
+        borderBottom: 'none',
+        padding: '0 10px',
+    },
+}))
+
+const addressList = [
```
Comment (nuanyang233) [inline packages/maskbook/src/plugins/Wallet/UI/ImportWalletDialog/ImportResult.tsx:29]: Do these look like demo data 👀, or do they have any special meaning?

File: packages/maskbook/src/plugins/Wallet/UI/ImportWalletDialog/MnemonicTab.tsx [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+interface MnemonicTabProps {
+    words: string[]
+    onChange: (words: string[]) => void
+}
+
+export const MnemonicTab: FC<MnemonicTabProps> = ({ words, onChange }) => {
+    const classes = useStyles()
+
+    const handleChange = (newWord: string, index: number) => {
+        const newWords = words.map((word, i) => {
```
Comment (nuanyang233) [inline packages/maskbook/src/plugins/Wallet/UI/ImportWalletDialog/MnemonicTab.tsx:44]: Perhaps using Map seems to perform better

File: packages/maskbook/src/plugins/Wallet/UI/ImportWalletDialog/FromJson.tsx (hunk 9 above)
Comment (UncleBill) [reply packages/maskbook/src/plugins/Wallet/UI/ImportWalletDialog/FromJson.tsx:4]: It's a placeholder. Should integrate the ImportFromJson component

File: packages/maskbook/src/plugins/Wallet/UI/ImportWalletDialog/FromPrivateKey.tsx (hunk 10 above)
Comment (UncleBill) [reply packages/maskbook/src/plugins/Wallet/UI/ImportWalletDialog/FromPrivateKey.tsx:66]: Might patched in another PR.

File: packages/maskbook/src/plugins/Wallet/UI/ImportWalletDialog/ImportResult.tsx (hunk 11 above)
Comment (UncleBill) [reply packages/maskbook/src/plugins/Wallet/UI/ImportWalletDialog/ImportResult.tsx:68]: ditto

File: packages/maskbook/src/plugins/Wallet/UI/ConnectWalletDialog/index.tsx (hunk 6 above)
Comment (UncleBill) [reply packages/maskbook/src/plugins/Wallet/UI/ConnectWalletDialog/index.tsx:89]: I think it's kind of waiting at least 1s to reduce flicker.

File: packages/maskbook/src/plugins/Wallet/UI/ImportWalletDialog/FromJson.tsx (hunk 9 above)
Comment (Jack-Works) [reply packages/maskbook/src/plugins/Wallet/UI/ImportWalletDialog/FromJson.tsx:4]: So this is a WIP?

File: pnpm-lock.yaml [hunk 14]
```diff
@@ -24219,4 +24219,4 @@ packages:
       shlex: 2.1.0
       slash: 2.0.0
       tmp: 0.0.33
-    dev: true
+    dev: true
```
Comment (Jack-Works) [inline pnpm-lock.yaml:24222]: 🤔?

File: packages/shared/src/hooks/useValueRef.ts [hunk 15]
```diff
@@ -12,3 +13,19 @@ export function useValueRef<T>(ref: ValueRef<T>) {
 
     return useSubscription(subscription)
 }
+
+export function useValueRefDelayed<T>(ref: ValueRef<T>, latency = 500) {
```
Comment (Jack-Works) [inline packages/shared/src/hooks/useValueRef.ts:17]: Can you describe what is this used for? I know the functionality of this but I don't know the context of using this.

**Comment by guanbinrui** [comment]
Replaced by #3269.

**Pull request closed without merging**